r1cs/gadgets/std/src/fields/nonnative/nonnative_field_gadget.rs

```diff
@@ -955,34 +1077,34 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         //      limb[i] < 2^surfeit * 2^bits_per_limb,
         // ``
         // and similarly for the most significant limb.
-        // Costs `num_limbs` many constraints.
+        // Costs no constraints.
         let mut kp_gadget_limbs = Vec::new();
-        for (i, limb) in p_gadget.limbs.iter().enumerate() {
-            let mul = limb.mul(cs.ns(|| format!("limb_{} * k_gadget", i)), &k_gadget)?;
+        for (i, limb) in p_representations.iter().enumerate() {
+            let mul = k_gadget.mul_by_constant(cs.ns(|| format!("limb_{} * k_gadget", i)), limb)?;
             kp_gadget_limbs.push(mul);
         }
 
-        // Enforce `delta = k*p` as big integers. Note that the surfeit of `k*p` is 
+        // Enforce `delta = k*p` as big integers. Note that the surfeit of `k*p` is
         // `1 + log(1 + num_adds(D))`, which is larger than the surfeit of `delta`.
         // Costs
         // ``
-        //  (S-1) * (1 + limb_size + 2 - shift_per_limb) + 1 =
-        //      (S-1) * (3 + surfeit) + 1
-        // ``             
-        // constraints, where
-        // `` 
-        // S - 1 = Floor[
-        //          (ConstraintF::CAPACITY - 2 - (bits_per_limb + surfeit)) / bits_per_limb
-        //      ] 
-        //      =  1 + Floor[
+        //      (num_groups - 1) * (1 + bits_per_limb + surfeit + 3 - shift_per_limb) + 2 =
+        //      (num_groups - 1) * (1 + surfeit + 3) + 2
+        // ``
+        // constraints, where `1 <= num_groups <= num_limbs` is the number of groups, determined
+        // by `num_groups = Ceil[num_limbs / S]` with
+        // ``
+        //  S - 1 = Floor[
+        //          (ConstraintF::CAPACITY - 2 - (bits_per_limb + surfeit)) / shift_per_limb
+        //      ] = Floor[
         //          (ConstraintF::CAPACITY - 2 - surfeit) / bits_per_limb
-        //      ],
+        //      ] - 1.
         // ``
```

Comment:
I don't understand why these two last expressions are equal.
According to your definition now S = num_limbs_in_a_group. We have that each limb is < 2^(bits_per_limb + surfeit).
We must be sure that 
2^(bits_per_limb + surfeit) * (1 + 2^shift_per_limb + ... + 2^(shift_per_limb * (S-1)) <= 2^(CAPACITY - 1).
We have that (1 + 2^shift_per_limb + ... + 2^(shift_per_limb * (S-1)) < 2 * 2^(shift_per_limb * (S-1)).
So the first S is the maximum one which achieves this inequality, but bits_per_limb != shift_per_limb here.

Reply:
I simplified since in the following call of `group_and_check_equality` bits_per_limb == shift_per_limb